### ISSUE TITLE
Should IHttpClientBuilder.AddTypedClient<TClient> register the TClient as Scoped?

### ISSUE THREAD

**Issue by dotnetjunkie** [dotnet/runtime#36067]
The current implementation of the `HttpClientBuilderExtensions.AddTypedClient<TClient>` extension method of the `Microsoft.Extensions.Http` project registers the `TClient` with the `Transient` lifestyle. I think this is incorrect and would like drop this here for discussion.

With the introduction of ASP.NET Core v2.1, the new `IHttpClientFactory` interface and its related extension methods were added to address the problems with reuse of `HttpClient` instances, as discussed [here](https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/), [here](https://github.com/dotnet/corefx/issues/11224), and [here](https://byterot.blogspot.com/2016/07/singleton-httpclient-dns.html), as [referenced](https://www.stevejgordon.co.uk/introduction-to-httpclientfactory-aspnetcore) by  Steve Gordon in his blog post on this matter.

**Problem being that HttpClients that are stored for the duration of the app domain can cause several issues, _most obvious one being that a `Singleton` HttpClient doesn't respect DNS changes._ This, of course, is exactly why the new extension infrastructure is put in place.**

What strikes me, though, is that, although there is seemingly the observation that caching of `HttpClient` for the duration of the application is bad, `AddTypedClient` registers the `HttpClient` consumers as `Transient`, as can be seen in the `AddTypedClient` implementation:

``` c#
public static IHttpClientBuilder AddTypedClient<TClient>(this IHttpClientBuilder builder)
    where TClient : class
{
    ...
    builder.Services.AddTransient<TClient>(s =>
    {
        var httpClientFactory = s.GetRequiredService<IHttpClientFactory>();
        var httpClient = httpClientFactory.CreateClient(builder.Name);
        var typedClientFactory = s.GetRequiredService<ITypedHttpClientFactory<TClient>>();
        return typedClientFactory.CreateClient(httpClient);
    });

    return builder;
}

```

The HttpClientFactory sample project shows a usage of this method:

``` c#
services.AddHttpClient("github", c =>
  {
      c.BaseAddress = new Uri("https://api.github.com/");
  })
  .AddHttpMessageHandler(() => new RetryHandler())
  .AddTypedClient<GitHubClient>();//GitHubClient depends on HttpClient and gets the "github" client

services.AddSingleton<IMyService, MyService>(); // Depends on GitHubClient
```

I think the registration of `Transient` consumers is problematic, as `Transient`—in term of Microsoft.Extensions.DependencyInjection—means _stateless_; not _short lived_. This means that the container allows those _stateless_ consumers of `HttpClient` to be injected into `Singleton` consumers **without the container's Scope Verfication feature to get tripped**. For instance, in the case of the `Singleton` `MyService` registration shown above, it will keep the `GitHubClient` alive for the duration of the application, and `GitHubClient` will keep its `HttpClient` alive for the duration of the application. This, as stated above, could lead to missing DNS changes, **effectively causing the application to become corrupted in a way that can only be fixed with a restart of the application**.

The injection of `HttpClient` instances into `Singleton` consumers could, therefore, lead to the type of bugs that you so hard tried to prevent using this new design. Because of the way `AddHttpClient` is designed, no warning signs will be given. Because of the way DNS caching works, problems only appear after the application is deployed to production; hardly ever on the developer's machine, making it crucial to warn developers in an early stage about this issue.

I think the solution to this problem is, therefore, rather straightforward (although, admittedly, a breaking change). Change the `AddTypedClient` method to change its `TClient` as `Scoped` instead, **because in that case, MS.DI's Scope Validation feature will prevent the client to be resolved from a root scope**:

``` c#
public static IHttpClientBuilder AddTypedClient<TClient>(this IHttpClientBuilder builder)
    where TClient : class
{
    // Register client as Scoped, allowing Captive Dependencies being detected in case the
    // ServiceProvider is built using the validateScopes flag.
    builder.Services.AddScoped<TClient>(s =>
    {
        // same code as before
    });

    return builder;
}
```

Although changing the lifestyle from `Transient` to `Scoped` can be considered to be a breaking change, due to how Scope Validation is configured in ASP.NET Core applications, exceptions will only be thrown in development mode; not when the application is deployed.

**Comment by ericstj** [comment]
Given this fix is breaking I'm reluctant to take it this late in 5.0.  Let's consider this for 6.0.  One way to make it non-breaking is to add a new method that does the scoped registration, `AddTypedClientScoped`.  @davidfowl @maryamariyan @dotnet/ncl any thoughts on this?

**Comment by halter73** [comment]
I'm agree this is too risky to take for 5.0. Anyone who was resolving a typed client from a the root provider will be broken.

> For instance, in the case of the Singleton MyService registration shown above, it will keep the GitHubClient alive for the duration of the application, and GitHubClient will keep its HttpClient alive for the duration of the application.

Yep. And this is the scenario that would break if we make the typed client registrations scoped. What's wrong with a Signleton service keeping an HttpClient alive for the duration of the application? The IHttpClientFactory which owns all the HttpMessageHandlers and therefore the connection pool is already a singleton.

I don't think anyone mentioned IDisposable in this thread yet, but I wonder if that's really what's at the heart of this issue. Resolving transient services that happen to implement IDisposable from root provider _is_ very problematic, and a typed client could implement IDisposable leading to problems where the container itself, not the singleton service, keeps the typed client alive for the duration of the application, but that's not unique to typed clients. See #36491.

>  One way to make it non-breaking is to add a new method that does the scoped registration, `AddTypedClientScoped`.

I think this would just create more confusion over which method to call. I don't see the problem with typed clients being transient unless the type implements IDisposable. Not implementing IDisposable is the easy way to work around leaking the type clients. If there's some reason the type client needs to be disposed at the end of the request/scope, manually registering it as a scoped service should be easy enough even without a new extension method.

**Comment by EnricoMassone** [comment]
@halter73 just a question, above you said: 

> What's wrong with a Signleton service keeping an HttpClient alive for the duration of the application?

Based on my understanding, having a captive HttpClient keeped alive for the entire duration of the application could lead to **missing reaction to DNS updates**, as pointed out above by @nicklundin08 

This seems to be confirmed by [this Micorsoft docs on IHttpClientFactory](What's wrong with a Signleton service keeping an HttpClient alive for the duration of the application?). My understanding is that the HttpMessageHandler contained in the captive HttpClient is kept alive for the entire application lifetime, which causes the DNS change issue. Put in other word, the whole HTTP client story works fine if and only if each time a consumer needs an HttpClient instance it goes to the IHttpClientFactory and asks for a new instance. At that point, the factory is able to pick up an HttpMessageHandler for its pool and to return a new HttpClient using the pooled handler. 

Is my understanding correct ?

**Comment by wfurt** [comment]
As far as I can tell, the HttpClient will not pro-actively pick up new nodes resolving to same name as far as there are available connections. But it does not cache any DNS lookups e.g. when time comes to create new connection, it will do new DNS lookup (subject to OS caching) and it will connect to some IP returned by the lookup. You can use PooledConnectionLifetime to force discovery of new nodes. It seems like the problem with DNS is valid only if given server stops serving content for given name _without_ closing the connection, right?

**Comment by EnricoMassone** [comment]
> As far as I can tell, the HttpClient will not pro-actively pick up new nodes resolving to same name as far as there are available connections. But it does not cache any DNS lookups e.g. when time comes to create new connection, it will do new DNS lookup (subject to OS caching) and it will connect to some IP returned by the lookup. You can use PooledConnectionLifetime to force discovery of new nodes. It seems like the problem with DNS is valid only if given server stops serving content for given name _without_ closing the connection, right?

Hi, 

I don't really know the underlying implementation details. My high level understanding is the following:

- the `HttpClient` delegates the HTTP request handling to the wrapped HttpMessageHandler
- the connections are handled at the HttpMessageHandler level 
- HttpMessageHandler must be pooled because creating an arbitrary amount of them could lead to socket exhaustion under heavy loads
- the HttpMessageHandler, by default, keeps the connections open indefinitely and this is the root cause of the DNS change issue

[This documentation](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests#httpclient-lifetimes) seems to confirm that, by default, HttpMessageHandler implementations keeps the connections open indefinitely, so that they are not able to react to DNS changes: 

> Pooling of handlers is desirable as each handler typically manages its own underlying HTTP connections; creating more handlers than necessary can result in connection delays. Some handlers also keep connections open indefinitely, which can prevent the handler from reacting to DNS changes

So basically the issue is related to connections not being closed, as you pointed out above.

**Comment by CarnaViire** [comment]
Triage:
1) We should not change the current lifetime as it will be a breaking change
2) We shouldn't add a new method for scoped registration because there's not enough benefit in doing it that would justify extending API surface
3) we should update documentation with two messages:
 - specify that the container will track IDisposable transient services so it is advised not to make your typed clients IDisposable, or you should dispose them yourself
 - note that HttpClient instances will hold the connection indefinitely (unless you specify the primary handler to be SocketsHttpHandler with PooledConnectionLifetime - applies to .NET Core). If you want to leverage HttpClientFactory recreating the connection, you would need to ask for a new instance of HttpClient (and also your Typed client)

**Comment by AroglDarthu** [comment]
4. How about just deprecating the existing `AddTypedClient` and adding a new extension method that takes an additional, non-optional parameter to specify the desired lifetime? Do this in v6 and drop the original in v7.

This issue pre-dates v5 by nearly two years and now it is being delayed to after v6?! Seriously?

**Comment by karelz** [comment]
@AroglDarthu 

> This issue pre-dates v5 by nearly two years and now it is being delayed to after v6?! Seriously?

Yes, it is old issue and we don't plan to prioritize it for 6.0 from these reasons:
- The issue is fairly complicated.
- It does not block too many customers.
- The scenarios where it matters are not super clear yet (something to follow up on here).
- There seem to be workarounds.
- Solution does not seem to be super-clear yet.
- Our team (NCL) is still ramping up on the component (which was transferred to us recently) together with learning about DI for the first time, so expertise is limited (especially on edge cases).
- There are 500+ other Networking issues to compete for our investment.

If the problem was more impactful, or easier to solve, we would be happy to prioritize it for 6.0. Currently that is not the case.

**Comment by CarnaViire** [comment]
I admit that it is not great that injecting a Typed client into a singleton is not checked automatically and it may result in a situation with loss of DNS changes. It also make things worse that it is not obvious which lifetime you will get when you register a Typed client via `AddHttpClient` method (without looking into docs of course) so it is difficult to be aware of lifetimes when injecting a Typed client.

However, if there are other unsolvable pains that we did not consider, please let us know. @AroglDarthu what exactly is the problem that you would try to solve by registering a Typed client as scoped?

JFYI. There are 16 overloads for registering a Typed client. Adding 16 more is A LOT. We would need a really good justification for that.

P.S.: `HttpClient`s produced by `HttpClientFactory` are transient-like objects, so for me it actually seems reasonable for Typed clients - that receive an instance of `HttpClient` in constructor - to have same lifetime as `HttpClient`.

**Comment by dotnetjunkie** [comment]
> Our team (NCL) is still ... learning about DI for the first time, so expertise is limited

This is very honest... but still quite disturbing.

**Comment by davidfowl** [comment]
I think generally we need to deprecate/remove handler rotation from the IHttpClientFactory. Refreshing DNS should be left up to the innermost handler implementation. That would simplify the lifetime semantics of the client factory and remove some of the confusion around when to use it.

**Comment by CarnaViire** [comment]
@davidfowl That's right and we are planning to do that. However there's still a case when a custom PrimaryHandler is supplied - and we might still want to leave rotation for these ones. We may have a discussion in  https://github.com/dotnet/runtime/issues/35987 if you wish.

**Comment by dotnetjunkie** [comment]
I raised this issue almost 6 years ago and 6 releases of .NET have passed. It would be good to get a definite answer from Microsoft on how this issue will be handled. I'm fine with "we accept the risk for our customers and leave the API as is," but I think it's important to not linger this any longer.
